arch/risc-v/src/mpfs/mpfs_usb.c

```diff
@@ -0,0 +1,3770 @@
+
+      if (privep->epstate == USB_EPSTATE_SENDING)
+        {
+          mpfs_req_complete(privep, -EPROTO);
+        }
+
+      memset((uint8_t *)&priv->ctrl, 0, sizeof(struct usb_ctrlreq_s));
+
+      usbtrace(TRACE_READ(USB_EPNO(EP0)), count0);
+
+      if (count0)
+        {
+          mpfs_read_rx_fifo((uint8_t *)&priv->ctrl, count0, EP0);
+        }
+
+      /* SETUP data is ready */
+
+      len = GETUINT16(priv->ctrl.len);
+      if (USB_REQ_ISOUT(priv->ctrl.type) && len > 0)
+        {
+          /* Yes.. then we have to wait for the OUT data phase to complete
+           * before processing the SETUP command.
+           */
+
+          privep->epstate = USB_EPSTATE_EP0DATAOUT;
+        }
+      else
+        {
+          /* This is an SETUP IN command (or a SETUP IN with no data). */
+
+          privep->epstate = USB_EPSTATE_IDLE;
+
+          /* Handle the SETUP OUT command now */
+
+          mpfs_ep0_setup(priv);
+        }
+
+      /* Ready for next setup data */
+
+      mpfs_ep0_ctrlread(priv);
+    }
+}
+
+/****************************************************************************
+ * Name: mpfs_usb_interrupt
+ *
+ * Description:
+ *   Handle the USB interrupt, for the device mode only.
+ *
+ * Input Parameters:
+ *   irq      - IRQ number (unused)
+ *   context  - Context (unused)
+ *   arg      - Pointer to the private data
+ *
+ * Returned Value:
+ *   OK unconditionally
+ *
+ ****************************************************************************/
+
+static int mpfs_usb_interrupt(int irq, void *context, void *arg)
+{
+  struct mpfs_usbdev_s *priv = (struct mpfs_usbdev_s *)arg;
+  uint16_t isr;
+  uint16_t pending_rx_ep;
+  uint16_t pending_tx_ep;
+  int i;
+
+  /* Get the device interrupts */
+
+  isr = getreg8(MPFS_USB_IRQ);
+
+  usbtrace(TRACE_INTDECODE(MPFS_TRACEINTID_INTERRUPT), isr);
+
+  /* Get the endpoint interrupts */
+
+  pending_tx_ep = getreg16(MPFS_USB_TX_IRQ);
+  pending_rx_ep = getreg16(MPFS_USB_RX_IRQ);
+
+  usbtrace(TRACE_INTDECODE(MPFS_TRACEINTID_EP_TX_IRQ), pending_tx_ep);
+  usbtrace(TRACE_INTDECODE(MPFS_TRACEINTID_EP_RX_IRQ), pending_rx_ep);
+
+  if (isr & RESET_IRQ_MASK)
+    {
+      /* Handle the reset */
+
+      mpfs_reset(priv);
+
+#ifdef CONFIG_USBDEV_DUALSPEED
+      priv->usbdev.speed = USB_SPEED_HIGH;
+      priv->usbdev.dualspeed = 0;
+#else
+      priv->usbdev.speed = USB_SPEED_FULL;
+      priv->usbdev.dualspeed = 1;
+#endif
+
+      return OK;
+    }
+
+  /* Serve Endpoint Interrupts first */
+
+  if (pending_tx_ep & 0x01)
+    {
+      mpfs_ctrl_ep_interrupt(priv, 0);
+    }
+
+  if (pending_tx_ep)
+    {
+      for (i = 1; i < MPFS_USB_NENDPOINTS; i++)
+        {
+          if ((pending_tx_ep & (1 << i)))
+            {
+              mpfs_ep_tx_interrupt(priv, i);
+            }
+        }
+    }
+
+  if (pending_rx_ep)
+    {
+      for (i = 0; i < MPFS_USB_NENDPOINTS; i++)
+        {
+          if ((pending_rx_ep & (1 << i)))
```

Comment:
```suggestion
          if ((pending_rx_ep & (1 << i)) != 0)
```
but I prefer type immune
```
if (((pending_rx_ep >> i) & 1) != 0)
```